### PR TITLE
Random state fixes

### PR DESCRIPTION
**Pull request by jgh9094**
[please review the [Contribution Guidelines](http://epistasislab.github.io/tpot/contributing/) prior to submitting your pull request. go ahead and delete this line if you've already reviewed said guidelines.]

## What does this PR do?

I changed the `random_state=#` variable for the cv/data splits to a user-specified random seed. This gives the user more control over reproducibility, as far as the splitting of data goes.

## Where should the reviewer start?

I have added a `seed`  argument to be passed into the constructor for both estimator classes. This is a good place to start. In both of these classes `random_state=#` has been set to `random_state=self.seed`. Additionally, the `self.seed` variable is also passed to the `objective_function_generator` due to it splitting data and using `random_state`. The `estimator_utils.py` files show how we pass the seed to the `objective_function_generator`.

## How should this PR be tested?

We can just run normal tests to make sure nothing breaks. This is only to increase user reproducibility for experiments. More will be added in the future in regard to this, but this is a good first step in that direction.

## Any background context you want to provide?

Previously, the `random_state` variable was set to a hardcoded number. This results in the data being split the same every time an experiment is done. I think by allowing users to specify a seed, we can increase the generality of our results while controlling the randomness better. 
 
## What are the relevant issues?

[you can link directly to issues by entering # then the number of the issue]

## Questions:

- Do the docs need to be updated?
 
Maybe? We can mention that a seed can be passed but will default to None if not.

- Does this PR add new (Python) dependencies?

No.

### CODE REVIEW

**Comment by perib** [comment]
This adds two seed parameters, `rng_seed` and `cv_seed`. I do not think that `cv_seed` is necessary and it may lead to confusion. 

If users wanted fix the cv seed while keeping the rest of the algorithm random, they can already provide an instance of an sklearn cv splitter object with the seed they desire.

for example

`cv = sklearn.model_selection.StratifiedKFold(n_splits=10, shuffle=True, random_state=42),`

**Comment by perib** [comment]
The functions in the configuration were designed to be compatible with the optuna trial class. This commit adds and `rng_` parameter which breaks compatibility with optuna. 

see more info: https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html

File: tpot2/tpot_estimator/estimator.py [hunk 1]
```diff
@@ -476,6 +499,20 @@ def __init__(self,  scorers,
...
+        self.rng = np.random.default_rng(rng_seed)
+        self.rng_seed = rng_seed
+
+        # set the numpy seed so anything using it will be consistent as well
+        # not sure if this is 100% true tho
+        np.random.seed(rng_seed)
+
+        # lock in a random seed if cv_seed is None
+        if cv_seed is None:
+            self.cv_seed = self.rng.integers(1000000)
```
Comment (perib) [inline tpot2/tpot_estimator/estimator.py:512]: This breaks sklearn conventions for estimators.

all hyper parameters must have a corresponding attribute that matches and is not changed. So in this case, self.cv_seed must equal cv_seed and should not be overwritten with the random value.

more information here: https://scikit-learn.org/stable/developers/develop.html

**Comment by perib** [comment]
It looks like these changes break tutorial 8. 

The API has changed to assume that the individual object has a mutation and crossover function that takes in an rng parameter. The [BaseIndividual](https://github.com/EpistasisLab/tpot2/blob/main/tpot2/individual_representations/individual.py#L10) class should be changed to reflect the new API. 

Alternatively, we could remove the BaseIndividual altogether, and have the mutation and crossover functions passed in separately. That would be a more functional programming approach. In PR #54, I added the option to pass in mutation and crossover functions (along with probabilities of each) separately. We would just have to make it clear in the documentation what the signature for that function needs to be (in this case, `mutate(individual, rng/random_state)`

**Comment by perib** [comment]
rather than passing an rng_seed to the GraphIndividual, it may be simpler to add a partial to the make_X_config_dictionary functions ([classifier version found here](https://github.com/EpistasisLab/tpot2/blob/97a316216c8c458ab94b53bbf7ec710785403a9d/tpot2/config/classifiers.py#L277C5-L277C38))

**Comment by perib** [comment]
Functions that take in a random_state parameter should have a default value of `=None` to still be usable without manually creating a value.

**Comment by jgh9094** [comment]
> This adds two seed parameters, `rng_seed` and `cv_seed`. I do not think that `cv_seed` is necessary and it may lead to confusion.
> 
> If users wanted fix the cv seed while keeping the rest of the algorithm random, they can already provide an instance of an sklearn cv splitter object with the seed they desire.
> 
> for example
> 
> `cv = sklearn.model_selection.StratifiedKFold(n_splits=10, shuffle=True, random_state=42),`
> 
> Other functions are also using a parameter called `rng_` or `rng_generator` that serves the same purpose.
> 
> for consistency with sklearn, I think we should have a single variable name, and call it `random_state`, in place of all the seed and rng variables
> 
> Sklearn also provides a [helpful function here for getting a np.random.randomstate instance](https://github.com/scikit-learn/scikit-learn/blob/2a2772a87b6c772dc3b8292bcffb990ce27515a8/sklearn/utils/validation.py#L1207)
> 
> [An example of it being used in the sklearn logistic regression class here.](https://github.com/scikit-learn/scikit-learn/blob/d99b728b3/sklearn/linear_model/_logistic.py#L792)

For consistency with sklearn, I think we should just have `random_state` being passed as an argument into the estimator instead of `rng_seed` and `cv_seed`. This would make it easier to understand what it is doing and being passed to, given that a lot of things in sklearn can get a `random_state` argument.

However, I don't think that the `random_state` argument is sufficient to get rid of the numpy.random.Generator object we create and pass around between classes during a run. The purpose of the `rng_` is to centralize the object that is responsible for the randomness that goes on throughout a run, in regard to the evolvers/mutation/crossover calls.

**Comment by jgh9094** [comment]
> The functions in the configuration were designed to be compatible with the optuna trial class. This commit adds and `rng_` parameter which breaks compatibility with optuna.
> 
> see more info: https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html
> 
> A solution could be to put hyperparametersuggestor.py into a class, and give it a seed parameter. Then the function signatures could remain the same, but use the provided seed/random state to generate the new hyperparameters.

This should be a point of discussion between the developers on what to prioritize. I understand that optuna would be a great addition to have, but reproducibility is another important aspect to consider. 

The idea of passing the seed to the hyperparameters is one we could implement/test. My only concern is having more than one random number generator instance throughout a run. This could make things harder to debug and make things less reproducible.

File: tpot2/tpot_estimator/estimator.py (hunk 1 above)
Comment (jgh9094) [reply tpot2/tpot_estimator/estimator.py:512]: Agreed.

**Comment by jgh9094** [comment]
> It looks like these changes break tutorial 8.
> 
> The API has changed to assume that the individual object has a mutation and crossover function that takes in an rng parameter. The [BaseIndividual](https://github.com/EpistasisLab/tpot2/blob/main/tpot2/individual_representations/individual.py#L10) class should be changed to reflect the new API.
> 
> Alternatively, we could remove the BaseIndividual altogether, and have the mutation and crossover functions passed in separately. That would be a more functional programming approach. In PR #54, I added the option to pass in mutation and crossover functions (along with probabilities of each) separately. We would just have to make it clear in the documentation what the signature for that function needs to be (in this case, `mutate(individual, rng/random_state)`
> 
> another option which may or may not work could be to keep have the random state be an attribute of the individual rather than parameter to each mutation function. These would be set during the generation of the initial population. When generating new individuals/functions, it would be on the user to decide whether or not to implement random_state, and the evolver should still work even if the user does not use it. This has the advantage of simplifying the evolver API

This may not be relevant to the topic of this discussion, but I think the mutation/crossover functionality should be separated from the individual. It makes sense to keep it there if we want to have each individual implement its own unique mutation/crossover operations, but I'm not sure how practical that would be. That being said, I think these calls make more sense to happen all within the population class, or having a `variation.py` class that takes in individual(s) and returns an offspring (i.e. offpsirng = mutate(parent) or offspring1, offspring2 = crossover(parent1, parent2)). The variation.py would handle all of the offspring generation and hold the function within the individuals.

**Comment by jgh9094** [comment]
> rather than passing an rng_seed to the GraphIndividual, it may be simpler to add a partial to the make_X_config_dictionary functions ([classifier version found here](https://github.com/EpistasisLab/tpot2/blob/97a316216c8c458ab94b53bbf7ec710785403a9d/tpot2/config/classifiers.py#L277C5-L277C38))

Something along the lines of: 

```
make_classifier_config_dictionary(n_samples=10, n_classes=None, random_state=None):
   .
   .
   .
    return {
            LogisticRegression: partial(params_LogisticRegression, random_state=random_state),
```

**Comment by jgh9094** [comment]
> Functions that take in a random_state parameter should have a default value of `=None` to still be usable without manually creating a value.

Can you be a little more specific or give an example of where this should happen? By not defaulting to None for a random number generator, we are guaranteed to only be using one generator, regardless if it was created with None or a seed (e.g. `rng = np.random.default_rng(seed)` and `rng = np.random.default_rng(None)`).

**Comment by perib** [comment]
> However, I don't think that the random_state argument is sufficient to get rid of the numpy.random.Generator object we create and pass around between classes during a run. The purpose of the rng_ is to centralize the object that is responsible for the randomness that goes on throughout a run, in regard to the evolvers/mutation/crossover calls.

What I meant was that they should all have the same variable name for consistency. 

>Can you be a little more specific or give an example of where this should happen? By not defaulting to None for a random number generator, we are guaranteed to only be using one generator, regardless if it was created with None or a seed (e.g. rng = np.random.default_rng(seed) and rng = np.random.default_rng(None)).

Currently, the selection functions are defined like this: 

`def tournament_selection(scores, k, rng_, n_parents=1, tournament_size=2, score_index=0)`

And I suggest they look like this:

`def tournament_selection(scores, k, rng_=None, n_parents=1, tournament_size=2, score_index=0)`


This gives the user the option of using the function with an numpy generator that they are passing around, or simply applying the function to the scores directly without needing to worry about creating a generator or setting the seed.

**Comment by perib** [comment]
> > rather than passing an rng_seed to the GraphIndividual, it may be simpler to add a partial to the make_X_config_dictionary functions ([classifier version found here](https://github.com/EpistasisLab/tpot2/blob/97a316216c8c458ab94b53bbf7ec710785403a9d/tpot2/config/classifiers.py#L277C5-L277C38))
> 
> Something along the lines of:
> 
> ```
> make_classifier_config_dictionary(n_samples=10, n_classes=None, random_state=None):
>    .
>    .
>    .
>     return {
>             LogisticRegression: partial(params_LogisticRegression, random_state=random_state),
> ```

yup!

**Comment by perib** [comment]
> The idea of passing the seed to the hyperparameters is one we could implement/test. My only concern is having more than one random number generator instance throughout a run. This could make things harder to debug and make things less reproducible.

I think there could be a way to do with with a single number generator too. 

we could change the config file format into something more like the TPOT1 dictionary or JSON files, which can then be passed into special functions for either optuna or the graph individual. 

or create a trial class that holds onto the same random number generator like `Trial(rng_=rng).suggest_categorical()` 

This is related to issue #78

**Comment by jgh9094** [comment]
> > The idea of passing the seed to the hyperparameters is one we could implement/test. My only concern is having more than one random number generator instance throughout a run. This could make things harder to debug and make things less reproducible.
> 
> I think there could be a way to do with with a single number generator too.
> 
> we could change the config file format into something more like the TPOT1 dictionary or JSON files, which can then be passed into special functions for either optuna or the graph individual.
> 
> or create a trial class that holds onto the same random number generator like `Trial(rng_=rng).suggest_categorical()`
> 
> This is related to issue #78

I'm not to familiar with how things are implemented in tpot1 so I can't really comment on that.

I think the second option is a good idea, but by having Trial() hold onto a rng, we will now have two maintained throughout a run. This would be way easier if we had C++ pointers...

**Comment by perib** [comment]
> I'm not to familiar with how things are implemented in tpot1 so I can't really comment on that.

The disadvantage of the TPOT1 option is that it doesn't allow for conditionals. 

> I think the second option is a good idea, but by having Trial() hold onto a rng, we will now have two maintained throughout a run. This would be way easier if we had C++ pointers...

The trial attribute of rng would be the same object as passed in, so there would not necessarily be two different generators. `Trial(rng_=rng).suggest_categorical()` would be equivalent to `suggest_categorical(rng_=rng)`. We don't have to keep the instance of Trial either (so we avoid multiple classes holding onto the same object, which can get confusing...).

I feel like there's probably a simpler solution though?
